clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -229,7 +229,7 @@ public long connectionDelay(Node node, long now) {
      */
     @Override
     public boolean connectionFailed(Node node) {
-        return connectionStates.connectionState(node.idString()).equals(ConnectionState.DISCONNECTED);
+        return connectionStates.isDisconnected(node.idString());
```

Comment:
To clarify: the issue is the `connectionFailed` method, changing the meaning of that method to `have not attempted a connection` is surprising and likely to lead to bugs in the future.

Another option is to use the result of the `poll` method. Each `ClientResponse` has a `disconnected` boolean and we could iterate over the responses instead of the unsent nodes. Is there any reason why this would not work?